SI-5906 Search for sorted sequences

alexclare opened the pull request:
Already reviewed as pull request #889.

alexclare committed:
SI-5906 Search for sorted sequences

Augments sequence classes with search functionality, using binary
search (comparable to that found in java.util.Collections) for
indexed sequences and linear search for others.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/529/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/529/

retronym commented on src/library/scala/collection/Searching.scala:
```diff
@@ -0,0 +1,100 @@
+    private def binarySearch[B >: A](elem: B, from: Int, to: Int)
+    (implicit ord: Ordering[B]): Either[Int, Int] = {
+      if ((to-from) == 1) Left(from) else {
+        val idx = (to+from)/2
```
Susceptible to overflow; see http://googleresearch.blogspot.ch/2006/06/extra-extra-read-all-about-it-nearly.html

retronym commented on src/library/scala/collection/Searching.scala:
```diff
@@ -0,0 +1,100 @@
+      * @param elem the element to find.
+      * @param ord  the ordering to be used to compare elements.
+      * @return a `Right` value containing the index corresponding to the element in the
+      *         $coll, or a `Left` value containing the index where the element would be
```
```
s/$coll/`coll`/
```

retronym commented on src/library/scala/collection/Searching.scala:
```diff
@@ -0,0 +1,100 @@
+      */
+    final def search[B >: A](elem: B)(implicit ord: Ordering[B]): Either[Int, Int] = 
+      coll match {
+        case _: IndexedSeqLike[A, Repr] => binarySearch(elem, -1, coll.length)(ord)
```
Why not `case _: IndexedSeq[A] =>` ?

alexclare replied:
I'd been using SeqLike for the rest of the section, so I chose IndexedSeqLike; was this a poor style choice?

retronym replied:
`IndexedSeq` is more general; one could write one that wasn't based on the `IndexedSeqLike` template trait.

retronym commented on src/library/scala/collection/Searching.scala:
```diff
@@ -0,0 +1,100 @@
+      * @return a `Right` value containing the index corresponding to the element in the
+      *         $coll, or a `Left` value containing the index where the element would be
+      *         inserted if the element is not in the $coll.
+      */
```
The implementation detail of choosing binary vs linear search based on the collection type is pretty important to document.

retronym commented on src/library/scala/collection/Searching.scala:
```diff
@@ -0,0 +1,100 @@
+      *         $coll, or a `Left` value containing the index where the element would be
+      *         inserted if the element is not in the $coll.
+      */
+    final def search[B >: A](elem: B)(implicit ord: Ordering[B]): Either[Int, Int] = 
```
Maybe create a data type for the result here.

```
sealed abstract class SearchResult {
  def insertionPoint: Int
}

case class Found(foundIndex: Int) extends SearchResult {
  def insertionPoint = foundIndex 
} 
case class InsertionPoint(insertionPoint: Int) extends SearchResult
```

retronym commented on src/library/scala/collection/Searching.scala:
```diff
@@ -0,0 +1,100 @@
+  *    import scala.collection.Searching._
+  *    val l = List(1, 2, 3, 4, 5)
+  *    l.search(3)
+  *    // == 2
```
Wrong result here.

alexclare commented:
Thanks for the feedback! The idea of making a specialized data type for search results is really great-it fixes @paulp's problem with the `Either[Int,Int]` return value.

alexclare committed:
Changes suggested by @retronym and @jsuereth

Change return type to case classes, branch between functions depending
on IndexedSeq instead of IndexedSeqLike, and alter tests accordingly.
Clean up doc comments and reflect changes in them.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/555/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/555/
